include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h

```diff
@@ -108,23 +118,244 @@ __pattern_transform_reduce(_ExecutionPolicy&& __exec, _ForwardIterator __first,
-
 template <typename T>
 struct ExecutionPolicyWrapper;
 
-template <typename _ExecutionPolicy, typename _Iterator1, typename _Iterator2, typename _UnaryOperation,
+template <typename _ExecutionPolicy, typename _InRng, typename _OutRng, typename _UnaryOperation,
           typename _InitType, typename _BinaryOperation, typename _Inclusive>
 void
-__pattern_transform_scan_base_impl(_ExecutionPolicy&& __exec, _Iterator1 __first, _Iterator1 __last,
-                                   _Iterator2 __result, _UnaryOperation __unary_op, _InitType __init,
-                                   _BinaryOperation __binary_op, _Inclusive)
+__pattern_transform_scan_single_group(_ExecutionPolicy&& __exec, _InRng&& __in_rng, _OutRng __out_rng, ::std::size_t __n,
+                              _UnaryOperation __unary_op, _InitType __init, _BinaryOperation __binary_op, _Inclusive)
 {
-    const auto __n = __last - __first;
-    assert(__n > 0);
+    using _CustomName = typename std::decay_t<_ExecutionPolicy>::kernel_name;
+    using _DynamicGroupScanKernel =
+                 oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<__par_backend_hetero::__scan_single_wg_dynamic_kernel, _CustomName, _BinaryOperation, _InRng, _OutRng>;
 
+    ::std::size_t __max_wg_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+
+#if _ONEDPL_COMPILE_KERNEL
+    auto __kernel = __par_backend_hetero::__internal::__kernel_compiler<_DynamicGroupScanKernel>::__compile(::std::forward<_ExecutionPolicy>(__exec));
+    __max_wg_size = ::std::min(__max_wg_size, oneapi::dpl::__internal::__kernel_work_group_size(
+                                                          ::std::forward<_ExecutionPolicy>(__exec), __kernel));
+#endif
+
+    // Specialization for devices that have a max work-group size of 1024
+    constexpr ::std::uint16_t __targeted_wg_size = 1024;
+
+    if (__max_wg_size >= __targeted_wg_size)
+    {
+        auto __single_group_scan_f = [&](auto __size_constant) {
+            constexpr ::std::uint16_t __size = decltype(__size_constant)::value;
+            constexpr ::std::uint16_t __wg_size = ::std::min(__size, __targeted_wg_size);
+            constexpr ::std::uint16_t __num_elems_per_item = __par_backend_hetero::__ceiling_div(__size, __wg_size);
+            const bool __is_full_group = __n == __wg_size;
+
+            if (__is_full_group)
+                __single_group_scan<_Inclusive::value, __num_elems_per_item, __wg_size, /* _IsFullGroup= */ true>::__launch_static_bounds_scan(__exec, __in_rng.all_view(), __out_rng.all_view(), __n, __init, __binary_op, __unary_op);
+            else
+                __single_group_scan<_Inclusive::value, __num_elems_per_item, __wg_size, /* _IsFullGroup= */ false>::__launch_static_bounds_scan(__exec, __in_rng.all_view(), __out_rng.all_view(), __n, __init, __binary_op, __unary_op);
+        };
+        if (__n <= 16)
+            __single_group_scan_f(std::integral_constant<::std::uint16_t, 16>{});
```

Comment:
`std::integral_constant<::std::uint16_t, 16>` Why it is so complicated ? 
For passing compile time constant  - enough using a template parameter, like this `__single_group_scan<16>(....)`

Reply:
This would work if we were able to use C++20's template lambdas since we can explicitly capture the compile-time integer as template parameter of the lambda, such as:
```c++
auto __single_group_scan_f = [&]<::std::uint16_t _Size>() { /* ... */ };
```

However, since we are limited to C++17, we have to either
- pass an argument to the lambda whose type embeds the integer, or
- use a free-standing function that accepts an integer template parameter, but all of the arguments that would have been captured by the lambda would have to be repeated on every line

Of these options, I believe the first one is cleaner.

Reply:
you can define a simple functor instead of lambda:
```
template<uint16_t __size, uint16_t __targeted_wg_size, uint16_t __wg_size>
struct __single_group_scan_f {
      constexpr ::std::uint16_t __wg_size = ::std::min(__size, __targeted_wg_size);
      constexpr ::std::uint16_t __num_elems_per_item = __par_backend_hetero::__ceiling_div(__size, __wg_size);
      auto operator()() {
      }
};
```